File: crates/recast-captions/src/highlight.rs
use crate::model::{CaptionAnimation, CaptionStyle, TranscriptWord};

/// Words considered spoken at source-time `t`. A word counts from the moment
/// `t` reaches its `start`, matching how ASS `\k` flips a syllable at its
/// boundary, and it stays counted, so a gap never un-highlights an earlier word.
pub fn spoken_word_count(words: &[TranscriptWord], t: f64) -> usize {
    let mut n = 0;
    for (i, w) in words.iter().enumerate() {
        if t >= w.start {
            n = i + 1;
        } else {
            break;
        }
    }
    n
}

/// Per-word centisecond durations for ASS `\k`, in chunk order.
///
/// Rounds the CUMULATIVE boundary and diffs consecutive boundaries, so the sum
/// equals the rounded total exactly; rounding each word on its own accumulates
/// drift across a line. Each word holds until the next one starts.
pub fn karaoke_centiseconds(words: &[TranscriptWord], chunk_start: f64) -> Vec<i64> {
    let cs = |s: f64| ((s - chunk_start) * 100.0).round() as i64;
    let mut durations = Vec::with_capacity(words.len());
    let mut previous = 0i64;
    for (i, w) in words.iter().enumerate() {
        let next_start = words.get(i + 1).map(|n| n.start).unwrap_or(w.end);
        let boundary = previous.max(cs(next_start));
        durations.push(boundary - previous);
        previous = boundary;
    }
    durations
}

/// Hex colour for a word: the active word wins the accent under `color`
/// emphasis, otherwise progressive highlight paints spoken words base and
/// unspoken muted, and `none`/`active` paint every word the base colour.
pub fn word_color<'a>(
    index: usize,
    active: Option<usize>,
    spoken: usize,
    anim: &'a CaptionAnimation,
    style: &'a CaptionStyle,
) -> &'a str {
    if Some(index) == active && anim.emphasis == "color" {
        return &anim.emphasis_color;
    }
    if anim.highlight() == "progressive" {
        return if index < spoken {
            &style.color
        } else {
            &style.muted_color
        };
    }
    &style.color
}

/// Whether word `index` scales up. Suppressed for a lone word, whose pop
/// entrance already carries the emphasis.
pub fn word_scaled(
    index: usize,
    active: Option<usize>,
    word_count: usize,
    anim: &CaptionAnimation,
) -> bool {
    anim.emphasis == "scale" && Some(index) == active && word_count > 1
}

File: packages/captions/src/highlight.ts
/**
 * Progressive highlight: how many words of a chunk are "spoken" at time `t`.
 *
 * This is the Loom model. Every word of the chunk is visible; the first
 * `spokenWordCount` of them are painted in the base colour and the rest are
 * muted. As `t` advances the count only grows within the chunk, so a spoken
 * word stays bright. Mirrored in Rust as the karaoke `\k` boundary count.
 */

import type { TranscriptWord } from "./types";

/**
 * Count of words considered spoken at source-time `t`.
 *
 * A word counts as spoken once `t` reaches its `start` (reaching the start is
 * the moment it lights up, matching how `\k` flips a syllable at its boundary).
 * `holdGaps` does not change the count here: once a word starts it stays
 * counted regardless, so gaps never un-highlight an earlier word.
 *
 * Returns 0 before the first word starts and `words.length` once the last has.
 */
export function spokenWordCount(words: Pick<TranscriptWord, "start" | "end">[], t: number): number {
	let n = 0;
	for (let i = 0; i < words.length; i++) {
		if (t >= words[i].start) n = i + 1;
		else break;
	}
	return n;
}

/**
 * Per-word centisecond durations for ASS `\k` karaoke, in chunk order.
 *
 * ASS `\k` uses cumulative centiseconds, and rounding each word independently
 * accumulates drift across a line. So we round the CUMULATIVE boundary of each
 * word to centiseconds and diff consecutive boundaries; the sum then equals the
 * rounded total exactly. Each word holds until the NEXT word starts (matching
 * the preview's gap-hold), and the last word holds to its own end.
 *
 * `chunkStart` is subtracted so the first boundary is measured from the chunk's
 * own start. All times are seconds.
 */
export function karaokeCentiseconds(
	words: Pick<TranscriptWord, "start" | "end">[],
	chunkStart: number,
): number[] {
	if (words.length === 0) return [];
	const cs = (s: number) => Math.round((s - chunkStart) * 100);
	const durations: number[] = [];
	let prevBoundary = 0;
	for (let i = 0; i < words.length; i++) {
		const nextStart = i + 1 < words.length ? words[i + 1].start : words[i].end;
		const boundary = Math.max(prevBoundary, cs(nextStart));
		durations.push(boundary - prevBoundary);
		prevBoundary = boundary;
	}
	return durations;
}
